File: engine/types/src/types/message.rs
// Message domain types for the OrqaStudio engine.
//
// Defines structs and enums for chat messages within a session, including roles,
// content types, stream status, and full-text search results over message history.
// These types flow from the database through the Tauri IPC boundary to the frontend.

use serde::{Deserialize, Serialize};

/// A single message block within a session, persisted to the database.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: i64,
    pub session_id: i64,
    pub role: MessageRole,
    pub content_type: ContentType,
    pub content: Option<String>,
    pub tool_call_id: Option<String>,
    pub tool_name: Option<String>,
    pub tool_input: Option<String>,
    pub tool_is_error: bool,
    pub turn_index: i32,
    pub block_index: i32,
    pub stream_status: StreamStatus,
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    pub created_at: String,
}

/// The role of the message author in the conversation.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum MessageRole {
    User,
    Assistant,
    System,
}

/// The content type of a message block.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ContentType {
    Text,
    ToolUse,
    ToolResult,
    Thinking,
    Image,
}

/// The streaming state of a message block.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum StreamStatus {
    Pending,
    Complete,
    Error,
}

/// Alias for the message primary key type, used in cross-module references.
pub type MessageId = i64;

/// A full-text search result over message history.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub message_id: i64,
    pub session_id: i64,
    pub session_title: Option<String>,
    pub content: String,
    pub highlighted: String,
    pub rank: f64,
}

File: libs/types/src/message.ts
export interface Message {
	readonly id: number;
	readonly session_id: number;
	readonly role: MessageRole;
	readonly content_type: ContentType;
	readonly content: string | null;
	readonly tool_call_id: string | null;
	readonly tool_name: string | null;
	readonly tool_input: string | null;
	readonly tool_is_error: boolean;
	readonly turn_index: number;
	readonly block_index: number;
	readonly stream_status: StreamStatus;
	readonly input_tokens: number | null;
	readonly output_tokens: number | null;
	readonly created_at: string;
}

export type MessageRole = "user" | "assistant" | "system";
export type ContentType = "text" | "tool_use" | "tool_result" | "thinking" | "image";
export type StreamStatus = "pending" | "complete" | "error";
export type MessageId = number;

export interface SearchResult {
	readonly message_id: number;
	readonly session_id: number;
	readonly session_title: string | null;
	readonly content: string;
	readonly highlighted: string;
	readonly rank: number;
}
